### DISCUSSION TITLE
clarification on what pins can supply what currents safely

### DISCUSSION THREAD

**Discussion by whyameye** [dbuezas/lgt8fx#305] [Q&A]
There seems to be a lot of seemingly conflicting information online and in the datasheet about what current the pins of the MCU can and should be able to handle. Referring to [this datasheet](https://github.com/watterott/LGT8F328P-Testing/raw/master/LGT8FX8P_databook_v1.0.5-English.pdf):
  - page 75: "Most I/O of LGT8FX8P can drive currency up to 30mA,"
  - page 81: "Normal port is driven by 12mA"
  - pages 98-99 show switching the 6 high current pins from 12ma to 80ma.
  - in [this thread](https://github.com/dbuezas/lgt8fx/issues/21#issuecomment-657020605) the OP refers to the normal mode for pins being 12/30ma. I'm not sure how to interpret that.
  - later in that thread @dbuezas says driving the pins higher than 12ma in low current mode is unsafe, even though they measure delivery of 40ma when shorted.
  - [this website](https://wolles-elektronikkiste.de/en/lgt8f328p-lqfp32-boards) says the pins can be driven at 30ma

I'm deep into trying to understand flakey behavior on a project which currently requires many of the pins to deliver 20ma (160 ohm load at 3.3V) and I'm trying to determine if the load on the pins could be a factor here.

**Comment by dbuezas** [comment]
Drawing too much current from an individual pin can result in:
* Overheating of the output FETs for the pin and eventual damage
* Undervoltage on the pin due to its non zero resistance

I've never observed flakiness in behavior due to this as you describe it.

There is also a limitation on total current used among all pins, (i think it was 200mA), did you check that?

You could also check that the input voltage is not too close to the brown out voltage, and if you are powering it via usb remember there's a protection diode that lowers the voltage half a volt or so.

What flakiness do you observe? Resets or fluctuating gpio voltage?

In any case, 30mA per pin is probably too high if you want reliability.

**Reply by whyameye** [reply]
Thanks. The flakiness I observe:

1. 1 unit started to reset itself randomly regardless of the code loaded into it
2. I2c communication is not reliable on all units
3. 1 unit seemed to show memory corruption: 1 variable would have 1 value in one part of the code and then another value within the same function without having been assigned a new value. Random unrelated changes in the code would affect this behavior.

I'm tried different power sources, including a bench power supply, USB and 3.7V battery. The MCU is running at 3.3V. I've monitored Vcc with the oscilloscope and it looks stable. I've also monitored GPIO I/O and that looks consistent. Sample size is less than 10 units.

**Comment by grossmag01** [comment]
I'm working on a documentation of the differences bewtween LGT8Fx  and their Atmel / Microchip counterparts for some time now. Maybe the following citation helps here. 

"All high-current capable GPIOs are accessible on all three packages. However, when using the 
LGT8F328P/QFP32  on a typical Uno or Nano clone board,  PF[5] and PF[4] are not  present on the pinout and PF[1] is electrically combined with PD[1]/TXD, so possible interference with serial communication must be checked.  Both A and B compare outputs of timer/counters 0, 1 and 3  can be set to high current mode, enabling low impedance PWM and frequency/waveform output.

The following table shows the internal resistance in Ohm of the  LGT8F328P ports  at Vcc= 5V in comparison to  the Atmega168/328. 
GPIO type \ status . . . . . . . . . . . . . . . . . .1 - HIGH  (R Pin to VCC)  . . . . . 0 - LOW  (R Pin to GND)		
LGT8F328P  Standard . . . . . . . . . . . . . . . .typ. 12 Ω. . . . . . . . . . . . . . . . . . . . typ. 10 Ω
LGT8F328P High current . . . . . . . . . . . . .typ. 4,5 Ω . . . . . . . . . . . . . . . . . . . typ. 3,7 Ω
Atmega168/328. . . . . . . . . . . . . . . . . . . . typ. 25 Ω. . . . . . . . . . . . . . . . . . . . typ. 25 Ω

Taking these values into account, the limit of 12 resp. 80 mA cited in the datasheet  seem very conservative. However, while the Atmega will withstand a momentary external short-circuit  of one GPIO at 5V , since the peak current does not exceed 200 mA,  a short-circuit of one of the  LGT8F328P outputs, specially  in the high current mode, has the potential to destroy the chip, if the Vcc supply doesn‘t limit the current to safe values  of i.e. 500 mA.  Besides the danger of thermally overloading the chip,  high sink /source  currents  can distort the voltage distribution  between the functional units of the MCU, leading for instance to a shift  oft the H/L threshold for digital inputs and the offset of analog functions like ADC, DAC and comparators.  This is aggravated by the fact, that the QFP32 and the SSOP20 packages have  just one Vcc and one GND connection for all functions . (QFP48 has two of each for the digital functions and an additional AVcc and AGND for the analog part. ).  A conservative upper limit for the total sink/source current per package seems to be 200 mA.

**Reply by whyameye** [reply]
This is this very helpful. Thanks!

My project uses the QFP48 and at any one time a maximum of 8 pins are sharing a load totalling 80ma. (each set of 2 pins provides the sink and the source for 1 device and there are 4 devices total. Each device is 160 ohms.) It sounds like this should not be a problem based on what you are sharing? Also I am only using 1 Vcc and one GND connection currently and it sounds like it would be unnecessary but still prudent for me to also connect to the other Vcc and GND connection on the QFP48. Does that all sound correct to you?

**Reply by grossmag01** [reply]
This looks like you are using the GPIO pins as H bridge or similar , correct ?   At  5 V,  you would have 30 mA per load, which is well in the acceptable range.  Is the load  purely resistive ? If there is an inductive component by i.e.  a relay coil or similar, the switching induces voltage spikes, which could explain spurious behaviour of the MCU. 

If you use the QFP48 package, it is essential to connect all 3 GND pins to ground and the 2 Vcc and the AVcc to the supply voltage. The different pins are connected to different regions of the chip and must be on the same potential.  In the QFP32 and SOP20, these are connected internally . I'm not sure whether this solves your problem, but it's possible

**Reply by whyameye** [reply]
Yes I am using GPIO pins as H bridge. There is an inductive component as the pins are powering coils. I've been worried about flyback voltage and suspected that first when I started to see problems. My understanding is there will be no flyback voltage from the coils ~if both ends of the coil are floating after being charged~ if both ends of the coil are powered at GND after being energized so I always make sure to do that in the software. From my oscilloscope readings it appears this trick is preventing flyback but it's worth keeping this on the radar as a potential issue for sure.

> If you use the QFP48 package, it is essential to connect all 3 GND pins to ground and the 2 Vcc and the AVcc to the supply voltage. 

We have both Vccs and all 3 GNDs connected to each other already. AVcc is currently connected to a .1uF cap that connects to GND. I could try connecting that to Vcc as well.

How did you learn that the Vccs weren't connected internally? Is that something I missed in the datasheet or did you figure that out emperically?

**Reply by grossmag01** [reply]
The chinese datasheet is only marginally specific about the electric properties of the chip, in contrast to most western manufacturers, which have example application schematics and lots of diagrams to show power consumption and the like at various operating conditions.  Having to connect  all GND  and Vcc  pins of the QFP48 is not explicitly mentioned anywhere .  In an inofficial statement in a chinese electronics forum, a representative of LGT stated that the voltage difference between  Vcc and  AVcc should be below 0.1 V.  A separate cap for the AVcc pin is o.k. , as long as it's connected to Vcc via a resistor of less than 50 Ohms or a choke of i.e. 500 uH. If preventing digital noise  to interfere with A/D conversions is not necessary, AVcc is best connected to Vcc directly. 

Switching inductances is always tricky. If you keep both sides of the coils  connected  to GPIOs in either H or L  state, there should be no extreme voltage spikes.  You can improve the situation by  having an additional  resistor of i.e. 500 Ohms in parallel to each coil.  If switching speed is not of the essence,  using an  RC snubber  on each coil ( cap of 0.1 uF and resistor of  22 Ohms in series ) could also help.  I have used the LGT8F328P to directly drive the  30 Ohms coils of a stepper motor ( in high current mode of course ) without any problems. 

By the way, have you seen any improvement of your problems already ?

**Comment by dwillmore** [comment]
Ummm, no.  Turning off the high and low side switches at the same time will
not prevent EMF from inducing a voltage across the coil.  The voltage will
be present across the coil and will both overvoltage the high side and
undervoltage the low side.  If you're not seeing this on the scope, it may
be that your scope is just loading the circuit enough to supress it--or
your scope has insufficient bandwidth to see the peaks.  If you really want
to drive an inductive load directly with a uC, you may want to look into
adding some diodes to all of the driving pins to protect them from >Vcc and
<GND.  Two diodes/GPIO will do it. There are small diodes on the uC die,
but they're not meant for taking this much energy and will fail and that
can lead to other problems with your chips.

**Reply by whyameye** [reply]
Oops I misspoke. I don't turn both sides off. I power both sides of the coil to GND. I'll edit my earlier comment to reflect that.

**Comment by LaZsolt** [comment]
Protecting a H bridge is not a simple task. Look at this complex protection: https://i.stack.imgur.com/aVFr3.jpg
>  If you really want to drive an inductive load directly with a uC, you may want to look into adding some diodes to all of the driving pins to protect them

It is recommended that these be fast schottky diodes.

**Reply by LaZsolt** [reply]
And let's not forget the mechanical energy stored by the rotating mass. When braking, the elecric motor produces energy.

**Comment by dbuezas** [comment]
In case you can still modify the design, I had success with DRV8833, they are very cheap low voltage mosfet double h-bridges (2.7-10.8v, around half an ohm rds-on)

**Reply by whyameye** [reply]
no question at this pont I'm modifying the design! :)

**Comment by dwillmore** [comment]
Or a BDR6122TC if you like'em cheap and tiny.
